### PR TITLE
Reserve IOCBs for tool applications

### PR DESCRIPTION
**Pull request by denesb**
Artifact tests have been failing since the switch to the native nodetool, because ScyllaDB doesn't leave any IOCBs for tools. On some setups it will consume all of them and then nodetool and any other native app will refuse to start because it will fail to allocate IOCBs.
This PR fixes this by making use of the freshly introduced `--reserve-io-control-blocks` seastar option, to reserve IOCBs for tool applications. Since the `linux-aio` and `epoll` reactor backends require quite a bit of these, we enable the `io_uring` reactor backend and switch tools to use this backend instead. The `io_uring` reactor backend needs just 2 IOCBs to function, so the reserve of 10 IOCBs set up in this PR is good for running 5 tool applications in parallel, which should be more than enough.

Fixes: https://github.com/scylladb/scylladb/issues/19185

The problem this PR fixes has a manual workaround (and is rare to begin with), no backport needed.

### CODE REVIEW

**Comment by yaronkaikov** [comment]
> Artifact tests have been failing since the switch to the native nodetool, because ScyllaDB doesn't leave any IOCBs for tools. On some setups it will consume all of them and then nodetool and any other native app will refuse to start because it will fail to allocate IOCBs. This PR fixes this by making use of the freshly introduced `--reserve-io-control-blocks` seastar option, to reserve IOCBs for tool applications. Since the `linux-aio` and `epoll` reactor backends require quite a bit of these, we enable the `io_uring` reactor backend and switch tools to use this backend instead. The `io_uring` reactor backend needs just 2 IOCBs to function, so the reserve of 10 IOCBs set up in this PR is good for running 5 tool applications in parallel, which should be more than enough.
> 
> Fixes: #19185
> 
> I think we shouldn't backport this. Our artifact tests have been failing ever since the switch to native nodetool, but on the other hand we received no user complaint about this so far. The problem only affects non-root install, which is not commonly used.
> 

@dani-tweig @roydahan @mykaul Maybe we should drop `nonroot` artifacts only for `6.2` release? just for making sure the offline-installer is passing, so we will not get any other hidden issues

> Depends on [scylladb/seastar#2533](https://github.com/scylladb/seastar/pull/2533) (CI will keep failing until it goes in).

**Comment by denesb** [comment]
If the failing nonroot tests are making us miserable, we can backport this just for that.

**Comment by denesb** [comment]
> ❌ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/232/)

docker artifact test failed with `07:28:52  sdcm.cluster.ClusterNodesNotReady: Not all nodes joined the cluster`. I checked the logs of the node and I see no error, the node is starting up fine. Maybe the timeout is too tight? @fruch

**Comment by fruch** [comment]
> > ❌ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/232/)
> 
> docker artifact test failed with `07:28:52  sdcm.cluster.ClusterNodesNotReady: Not all nodes joined the cluster`. I checked the logs of the node and I see no error, the node is starting up fine. Maybe the timeout is too tight? @fruch 

Check SCT log, nodetool is probably failing

**Comment by yaronkaikov** [comment]
@denesb I don't think we need this to be backported to 6.1, Offline-installer is passing :-)

**Comment by yaronkaikov** [comment]
Also added label for running artifact tests for this CI, This way we can make sure the fix is working before merging :-)

**Comment by fruch** [comment]
> > > ❌ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/232/)
> > 
> > 
> > docker artifact test failed with `07:28:52  sdcm.cluster.ClusterNodesNotReady: Not all nodes joined the cluster`. I checked the logs of the node and I see no error, the node is starting up fine. Maybe the timeout is too tight? @fruch
> 
> Check SCT log, nodetool is probably failing

@denesb 

yeah, nodetool was failing like the following:

```
< t:2024-11-15 10:41:29,815 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <172.17.0.2>: terminate called after throwing an instance of 'std::invalid_argument'
< t:2024-11-15 10:41:29,815 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <172.17.0.2>:   what():  find_candidate(): failed to find candidate io_uring
< t:2024-11-15 10:41:29,815 f:base.py         l:141  c:RemoteLibSSH2CmdRunner p:DEBUG > <172.17.0.2>: STDERR: terminate called after throwing an instance of 'std::invalid_argument'
< t:2024-11-15 10:41:29,815 f:base.py         l:141  c:RemoteLibSSH2CmdRunner p:DEBUG >   what():  find_candidate(): failed to find candidate io_uring
```

**Comment by yaronkaikov** [comment]
@denesb, @fruch  what is the next step here?

**Comment by denesb** [comment]
> @denesb, @fruch what is the next step here?

I need to debug this. I tried to reproduce locally, but it doesn't. And I have a hard time grasping that -- whether the `io_uring` is available or not is decided at build time (whether we include the code in the build). So this is not something that is env dependent. Could it be that the test uses a different (older) binary for nodetool?

I am currently busy with triaging tablet related issues, I will get back to this when I can.

**Comment by denesb** [comment]
I think the problem is that `io_uring` is not available in docker. I added a fallback to `epoll`, let's see how it goes.

**Comment by yaronkaikov** [comment]
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13321/) ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/334/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1832/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12942/) ✅ - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/12/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4929/) ❌ - [Rpm Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-centos-rpm/6/) ✅ - [Deb Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-unified-deb/5/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13321/)
> 
> #### Build Details:
> * Duration: 3 hr 51 min
> * Builder: spider8.cloudius-systems.com

@denesb The failure is not related to this PR. however, i have re-run the offline-installer artifacts with all tests and it still failing for the non-root jobs - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/14/

**Comment by yaronkaikov** [comment]
> > > #### 🔴 CI State: FAILURE
> > > ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13321/) ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/334/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1832/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12942/) ✅ - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/12/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4929/) ❌ - [Rpm Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-centos-rpm/6/) ✅ - [Deb Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-unified-deb/5/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13321/)
> > > #### Build Details:
> > > 
> > > * Duration: 3 hr 51 min
> > > * Builder: spider8.cloudius-systems.com
> > 
> > 
> > @denesb The failure is not related to this PR. however, i have re-run the offline-installer artifacts with all tests and it still failing for the non-root jobs - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/14/
> 
> @yaronkaikov you still need the change in SCT to point to the right nodetool command

@fruch after your merged https://github.com/scylladb/scylla-cluster-tests/pull/8803 the job still failing with
```
11:19:46  Command: '/home/scylla-test/scylladb/bin/nodetool  status '
11:19:46  
11:19:46  Exit code: 1
11:19:46  
11:19:46  Stdout:
11:19:46  
11:19:46  
11:19:46  
11:19:46  Stderr:
11:19:46  
11:19:46  ERROR 2024-11-24 09:19:28,970 seastar - Could not setup Async I/O: Resource temporarily unavailable. The required nr_event (1024) exceeds the limit of request capacity in /proc/sys/fs/aio-max-nr (65536). Try increasing that number or reducing the amount of logical CPUs available for your application
```
https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/artifacts-offline-install/job/artifacts-rocky9-nonroot-test/22/console

**Comment by denesb** [comment]
New in v3:
* Fall-back to epoll when io_uring is not available.
* Also adjust the IOCB reserves when io_uring is not available: epoll needs 1024 IOCBs, instead of 2 for io_uring.

**Comment by denesb** [comment]
> @denesb I don't think we need this to be backported to 6.1, Offline-installer is passing :-)

Is a 6.2 backport needed?

**Comment by yaronkaikov** [comment]
> > @denesb I don't think we need this to be backported to 6.1, Offline-installer is passing :-)
> 
> Is a 6.2 backport needed?

@mykaul @avikivity  what do you think?

**Comment by avikivity** [comment]
> > > @denesb I don't think we need this to be backported to 6.1, Offline-installer is passing :-)
> > 
> > 
> > Is a 6.2 backport needed?
> 
> @mykaul @avikivity what do you think?

Does this fix a critical bug or a regression?

How did 6.2 survive so far without it?

**Comment by denesb** [comment]
> > > > @denesb I don't think we need this to be backported to 6.1, Offline-installer is passing :-)
> > > 
> > > 
> > > Is a 6.2 backport needed?
> > 
> > 
> > @mykaul @avikivity what do you think?
> 
> Does this fix a critical bug or a regression?
> 
> How did 6.2 survive so far without it?

Not critical in any way. I added backport labels because the artifacts tests were failing for a long time, but I was just told that they are not failing on 6.1 anymore. So I asked about 6.2.

**Comment by avikivity** [comment]
> > > > > @denesb I don't think we need this to be backported to 6.1, Offline-installer is passing :-)
> > > > 
> > > > 
> > > > Is a 6.2 backport needed?
> > > 
> > > 
> > > @mykaul @avikivity what do you think?
> > 
> > 
> > Does this fix a critical bug or a regression?
> > How did 6.2 survive so far without it?
> 
> Not critical in any way. I added backport labels because the artifacts tests were failing for a long time, but I was just told that they are not failing on 6.1 anymore. So I asked about 6.2.

The artifact tests for 6.2 were working at some point. If they regressed, we should understand why.

**Comment by denesb** [comment]
> The artifact tests for 6.2 were working at some point. If they regressed, we should understand why.

Maybe we are not talking about the same tests.

@yaronkaikov @fruch what is the state of the artifact tests? I had the impression that they are failing ever since native nodetool was introduced. There is this really old issue for this which is still open: https://github.com/scylladb/scylladb/issues/19185.

**Comment by yaronkaikov** [comment]
> > The artifact tests for 6.2 were working at some point. If they regressed, we should understand why.
> 
> Maybe we are not talking about the same tests.
> 
> @yaronkaikov @fruch what is the state of the artifact tests? I had the impression that they are failing ever since native nodetool was introduced. There is this really old issue for this which is still open: #19185.

The artifacts for offline installer never passed for 6.2, the other artifacts are ok.

**Comment by fruch** [comment]
@denesb 
rocky8 artifact test, failed to start scylla:
```
Nov 27 05:47:18 artifacts-rocky8-jenkins-db-node-347315ab-0-1 scylla[79872]: FATAL: Exception during startup, aborting: std::invalid_argument (find_candidate(): failed to find candidate linux-aio)
```

**Comment by denesb** [comment]
New in v4:
* Don't unconditionally select `linux-aio` as default backend -- it may not be available. If this is the case, select `epoll` as the default backend. This is the same behaviour as when `io_uring` is not compiled in, i.e. how it was before this PR.

**Comment by denesb** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13411/) ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/391/) ✅ - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/16/) ❌ - [Rpm Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-centos-rpm/7/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1905/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5009/) ✅ - [Deb Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-unified-deb/6/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13023/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13411/)
> #### Build Details:
> 
>     * Duration: 3 hr 48 min
> 
>     * Builder: spider2.cloudius-systems.com

This is CI for v3, I posted v4 before it finished. Hopefully v4 will fix the reason this failed (unconditional use of `linux-aio` backend).

**Comment by yaronkaikov** [comment]
@denesb For offline installer we have only one failure now - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/20/ ,  https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/artifacts-offline-install/job/artifacts-rocky8-nonroot-test/25/

**Comment by denesb** [comment]
> @denesb For offline installer we have only one failure now - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/20/ , https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/artifacts-offline-install/job/artifacts-rocky8-nonroot-test/25/

I don't understand what went wrong here.

**Comment by fruch** [comment]
@yaronkaikov 

this rocky8-offline test seems to be broken and missing the `scylla-server` service 
since jmx was removed 

the diff when it broke: `git log 612a1416604..6d8e9645ce7d --oneline`

**Comment by fruch** [comment]
took it for a ride with offline nonroot runs:
- [ ] :clock1: https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-rocky8-test/17/
- [ ] :clock1: https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-rocky9-test/47/
- [ ] :clock1: https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-debian11-test/8/
- [ ] :clock1: https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-amazon2023-test/14/
- [ ] :clock1: https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-ubuntu2004-test/30/
- [ ] :clock1: https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-ubuntu2404-test/24/

**Comment by yaronkaikov** [comment]
> took it for a ride with offline nonroot runs:
> 
> * [ ]  🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-rocky8-test/17/
> * [ ]  🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-rocky9-test/47/
> * [ ]  🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-debian11-test/8/
> * [ ]  🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-amazon2023-test/14/
> * [ ]  🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-ubuntu2004-test/30/
> * [ ]  🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-ubuntu2404-test/24/

@fruch here - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/20/ we had all non-root runs as well (all passing except rocky8)

**Comment by fruch** [comment]
> > took it for a ride with offline nonroot runs:
> > 
> > * [ ]   🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-rocky8-test/17/
> > * [ ]   🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-rocky9-test/47/
> > * [ ]   🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-debian11-test/8/
> > * [ ]   🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-amazon2023-test/14/
> > * [ ]   🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-ubuntu2004-test/30/
> > * [ ]   🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-ubuntu2404-test/24/
> 
> @fruch here - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/20/ we had all non-root runs as well (all passing except rocky8)

it's not written clear on this PR was is getting tested..
anyhow you are raising an issue for rock8 offline nonroot ?

**Comment by yaronkaikov** [comment]
> > > took it for a ride with offline nonroot runs:
> > > 
> > > * [ ]    🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-rocky8-test/17/
> > > * [ ]    🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-rocky9-test/47/
> > > * [ ]    🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-debian11-test/8/
> > > * [ ]    🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-amazon2023-test/14/
> > > * [ ]    🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-ubuntu2004-test/30/
> > > * [ ]    🕐 https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-ubuntu2404-test/24/
> > 
> > 
> > @fruch here - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/20/ we had all non-root runs as well (all passing except rocky8)
> 
> it's not written clear on this PR was is getting tested.. anyhow you are raising an issue for rock8 offline nonroot ?

@syuu1228 already sent a fix   - https://github.com/scylladb/scylladb/pull/21721 - @denesb once this gets in, I assume you will get a Green CI :-)

**Comment by denesb** [comment]
> @syuu1228 already sent a fix - #21721 - @denesb once this gets in, I assume you will get a Green CI :-)

I already did, see https://github.com/scylladb/scylladb/pull/21527#issuecomment-2505694565.

**Comment by yaronkaikov** [comment]
> > @syuu1228 already sent a fix - #21721 - @denesb once this gets in, I assume you will get a Green CI :-)
> 
> I already did, see [#21527 (comment)](https://github.com/scylladb/scylladb/pull/21527#issuecomment-2505694565).


@denesb The default Offline-installer is running **without** non-root tests (since it was always failing), once we get a full offline-installer pass, we can enable it back to CI , to verify we don't have new regressions

**Comment by avikivity** [comment]
I don't understand the conversation past CI passing. Is there a problem or not?

**Comment by denesb** [comment]
Rebased.

**Comment by denesb** [comment]
> ❌ - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/23/)

```
10:03:17  java.io.StreamCorruptedException: invalid stream header: 636F7272
```
This time it seems to be infra issue. I will restart CI.

**Comment by yaronkaikov** [comment]
@denesb together with the last fix from @syuu1228 (https://github.com/scylladb/scylladb/pull/21747) all the offline installer tests are passing , I think this one is ready to get in :-)

**Comment by yaronkaikov** [comment]
@denesb are we waiting for someone to review it again? or just merge it?

File: main.cc [hunk 1]
```diff
@@ -685,6 +685,17 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
...
+        const auto candidates = app_cfg.reactor_opts.reactor_backend.get_candidate_names();
+
+        // We don't wan't ScyllaDB to run with the io_uring backend.
+        // So select the default reactor backend explicitely here.
+        if (std::ranges::find(candidates, "linux-aio") == candidates.end()) {
+            app_cfg.reactor_opts.reactor_backend.select_default_candidate("epoll");
+        } else {
+            app_cfg.reactor_opts.reactor_backend.select_default_candidate("linux-aio");
+        }
+    }
```
Comment (avikivity) [inline main.cc:707]: What happens if someone overrides with --reactor-backend=blah? Is it respected?

File: main.cc [hunk 2]
```diff
@@ -685,6 +685,17 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
 #endif
     app_cfg.auto_handle_sigint_sigterm = false;
     app_cfg.reactor_opts.max_networking_io_control_blocks.set_default_value(50000);
+    {
+        const auto candidates = app_cfg.reactor_opts.reactor_backend.get_candidate_names();
+
+        // We don't wan't ScyllaDB to run with the io_uring backend.
+        // So select the default reactor backend explicitely here.
+        if (std::ranges::find(candidates, "linux-aio") == candidates.end()) {
```
Comment (avikivity) [inline main.cc:693]: There is contains().

File: main.cc (hunk 1 above)
Comment (denesb) [reply main.cc:707]: Yes, we only set the default here. The user can override via the command-line.

**Comment by denesb** [comment]
New in v4:
* Rebased
* Use `std::ranges::contains()`

File: main.cc (hunk 2 above)
Comment (denesb) [reply main.cc]: Switched all use of `std::ranges::find()` to `std::ranges::contains()` in v4.

**Comment by denesb** [comment]
Artifacts tests failed with timeout, but I see both ScyllaDB and nodetool running fine.
@yaronkaikov @fruch what went wrong?

**Comment by denesb** [comment]
```
t:2024-12-04 11:19:51,192 f:remote_base.py  l:560  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.150>: Running command "cat /etc/scylla.
     d/cpuset.conf"...  
```

Looks like SSH to the node is what timed out.

**Comment by yaronkaikov** [comment]
Can we merge this?

**Comment by denesb** [comment]
@avikivity ping for merge.

**Pull request merged**